### ISSUE TITLE
Mapped array/tuple types can't be used as generic with a constraint

### ISSUE THREAD

**Issue by Jamesernator** [microsoft/TypeScript#48740]
# Bug Report

### 🔎 Search Terms

mapped tuple type

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about mapped type

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&ts=4.7.0-dev.20220417#code/MYGwhgzhAEBqYgCoE8AOBTaBvAUNf0AxAJYTxJqYC80ALgE4Cu6A3DgL46iQwCSAzACZo6AB610AOwAmMcigzY8BEhAHCaDZm07co0AKLj09SfXQAzEcZlyECzLgJFSRiafNXNTVhxxdwfQAZAHtgBAAeRGsJWzh7SgA+JWdzMGkQyRBkIlpKAC5oRDZlfGBMiC1gWhD6AAo8jELEAEoU5wJaAAtSADpCRuo6SjZnTlLoAHN0WmGMOpbm9o7oc1pGUzoeiH7B0YJOccHoUPCQCAAxWqjKGKlZaAAldHTM7IBBenowZAj5JOSNCcBAA9CCircAOQOADaAGkALqQ6AZdAwSQhWYQMC0UgWHLdTDlSSVb7ESSzSH-DCQ3oTMEEBzQaGUGEAa3QyBCVgcSOgpGgGNmPGIk0kYAARiBMDU5pgqQkaXTnAznEyWRgYaTyZMEdAAD4QzWSRgAWwlJj1hthEGQ5pCID5AqF0BFYsl0roITlzOp6Fp9PBK3VNoYOqd6MxrqgovFUpl3uOCooSrqgn4ABYMy0JvD+ZJoByuTzKAjCqdIrDEYkOCUGYhtvyYJViCAQFsibVzNU5QAaaAW8KMCCYU1gVAYaSur4-EHrVDS5u4ts4BkAK2HszHqA7xHo0AAbghmDBpGjUMQJB3XRKQgfMCZ6LUuBVZiAwggIOWP+crvQIjC6j9m4JhmJYCKAtAMITJI6AAO4nD+dSwQh6gLC0vYwfBiFnMh2EgR4ljoZhCIsEAA)

### 💻 Code

```ts
class ValType {
    #isValType = true;
}
class I32 extends ValType {
    #isI32 = true;
}
class Externref extends ValType {
    #isExternref = true;
}

class Local<T extends ValType> {
    readonly #type: T;

    constructor(type: T) {
        this.#type = type;
    }

    get type(): T {
        return this.#type;
    }
}

type LocalsFor<Type extends ReadonlyArray<ValType>> = {
    // Type 'Type[K]' does not satisfy the constraint 'ValType'.
    //   Type 'Type[keyof Type]' is not assignable to type 'ValType'.
    //     Type 'Type[string] | Type[number] | Type[symbol]' is not assignable to type 'ValType'.
    //       Type 'Type[string]' is not assignable to type 'ValType'.(2344)
    [K in keyof Type]: Local<Type[K]>
};

// This is still the correct type, because mapped array/tuples still
// just map their values despite the above error
const locals: LocalsFor<[I32, Externref]> = [
    new Local(new I32()),
    new Local(new Externref()),
];
```

### 🙁 Actual behavior

It produces an error in the mapped type that `Type[K]` doesn't satisfy the constraint.

### 🙂 Expected behavior

There should be no error, as mapped types do not map over all keys in the array/tuple, it should only consider `Type[K]` to be the numeric keys of the tuple, i.e. `Type[K]` in the mapped type will only ever be (subtypes of) `ValType` because ultimately `K` will only be assigned to the numeric keys of `Type` as it's an array type.

In this particular case, we have an explicit `Type extends ReadonlyArray<ValType>` so TypeScript should be able to deduce just fine that this particular mapped type will only ever be over an array, and hence should be able to deduce that `Type[K]` will only ever be `Type[number]` (or ``Type[`${ number }`]`` for tuples).

I [had another issue, however I closed it](https://github.com/microsoft/TypeScript/issues/48739) and opened this one because this is likely the fundamental cause of both problems.

**Comment by whzx5byb** [comment]
I think it is explained in https://github.com/microsoft/TypeScript/issues/48398.

For you example, a workaround is replace `Local<Type[K]>` with `Local<Type[K & number]>`

**Comment by Jamesernator** [comment]
> I think it is explained in #48398.

The fact tuple keys aren't subtypes of `number` is explained, however that is a separate issue.

> For you example, a workaround is replace `Local<Type[K]>` with `Local<Type[K & number]>`

No, that doesn't work, it causes the final type in the example to become:

```ts
const locals: [Local<I32 | Externref>, Local<I32 | Externref>]
```

Which is not correct, the existing `[Local<I32>, Local<Externref>]` is correct already actually, it's just that it warns in the mapped type despite the fact mapped types when applied to tuples/arrays just map over the entries.

**Comment by whzx5byb** [comment]
> No, that doesn't work

You are right. I believe it is necessary to introduce a generic type `KeyOf<T>` to provide only numeric values for array/tuple. It could be written as below.

```ts
type _GetKeysFromLength<
  L extends number,
  R extends readonly unknown[] = [],
> = number extends L ? number : R['length'] extends L ? R[number] : _GetKeysFromLength<L, readonly [R['length'], ...R]>;

type KeyOf<T extends unknown[]> = _GetKeysFromLength<T['length']> extends infer U extends number ? U : never;

type R1 = KeyOf<[1,2,3]>;
//   ^? type R1 = 0 | 2 | 1
type R2 = KeyOf<number[]>;
//   ^? type R2 = number


type LocalsFor<Type extends Array<ValType>> = {
    [K in KeyOf<Type>]: Local<Type[K]>
};

const locals: LocalsFor<[I32, Externref]> = [
    new Local(new I32()),
    new Local(new Externref()),
]; // seems OK now!

var l0 = locals[0]
// ^? var l0: Local<I32>
var l1 = locals[1]
// ^? var l1: Local<Externref>
```

**Comment by RyanCavanaugh** [comment]
This is a correct error. Matching the constraint ` extends ReadonlyArray<ValType>` does not provide a bound on an arbitrary `Type[keyof Type]`, as shown:
```ts
type M = ReadonlyArray<ValType> & { k: boolean };
type LocalsFor<Type extends ReadonlyArray<ValType>> = {
    // Type 'Type[K]' does not satisfy the constraint 'ValType'.
    //   Type 'Type[keyof Type]' is not assignable to type 'ValType'.
    //     Type 'Type[string] | Type[number] | Type[symbol]' is not assignable to type 'ValType'.
    //       Type 'Type[string]' is not assignable to type 'ValType'.(2344)
    [K in keyof Type]: Local<Type[K]>
};
type G = LocalsFor<M>;
type X = G["k"]; // x: Local<boolean>, a constraint violation
```

**Comment by Jamesernator** [comment]
> Matching the constraint ` extends ReadonlyArray<ValType>` does not provide a bound on an arbitrary `Type[keyof Type]`, as shown:

These bizarre rules make the whole point of "homomorphic mapped types" seem fairly weak and unreliable, like if there's gonna be special casing for arrays in mapped types to not map stuff like methods, why would we want it to map other random properties that might be tacked onto the array? (as an example, symbol-keyed protocols like `Symbol.for("nodejs.util.inspect.custom")` and such)

The amount of bizarre behaviours in the mapped types feature just utterly constantly surprises me, I don't see how any developers who aren't intimately familiar with TypeScript's internals are expected to have any reasoning about these complicated features (and it doesn't help that that [handbook doesn't even mention arrays are (kind've) homomorphic in mapped types](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html) to begin with).

**Comment by Jamesernator** [comment]
Either way, I think I'll reopen the other issue because that one at the very least checks the keys for "numeric"-ness.

**Comment by RyanCavanaugh** [comment]
> The amount of bizarre behaviours in the mapped types feature just constantly surprises me

Every single rule we've added to mapped types was put there to remove some *other* surprise 🤷‍♂️

**Comment by jcalz** [comment]
So is #27995 not really a bug then?

**Comment by jcalz** [comment]
This has changed from "working as intended" to "fixed" by #48837.  Not sure if anyone wants to alter tags here.
